src/waitress/channel.py

```diff
@@ -153,58 +162,70 @@ def handle_read(self):
         if data:
             self.last_activity = time.time()
             self.received(data)
+        else:
+            # Client disconnected.
+            self.connected = False
+
+    def send_continue(self):
+        """
+        Send a 100-Continue header to the client. This is either called from
+        receive (if no requests are running and the client expects it) or at
+        the end of service (if no more requests are queued and a request has
+        been read partially that expects it).
+        """
+        self.request.expect_continue = False
+        outbuf_payload = b"HTTP/1.1 100 Continue\r\n\r\n"
+        num_bytes = len(outbuf_payload)
+        self.outbufs[-1].append(outbuf_payload)
+        self.current_outbuf_count += num_bytes
+        self.total_outbufs_len += num_bytes
+        self.sent_continue = True
+        self._flush_some()
```

Comment:
I think this needs to use `_flush_some_if_lockable()` now that it may be invoked from the service thread, not just from the asyncore thread.

Reply:
The reason I think this is that the asyncore thread could still be writing outbufs even though there is no active request, so you have no guarantee right now that there's no writing going on when you invoke `_flush_some` in `send_continue` from the service thread.

Reply:
I am similarly concerned with appending data to outbufs without a lock. I think this entire block should be wrapped in `with self.outbuf_lock:` to avoid interfering with an active flush.

Reply:
Sorry for so many comments here. My proposal is to keep using _flush_some but to wrap the block in a lock:

```python
with self.outbuf_lock:
    self.outbufs[-1].append(outbuf_payload)
    self.current_outbuf_count += num_bytes
    self.total_outbufs_len += num_bytes
    self.sent_continue = True
    self._flush_some()
self.request.completed = False
```

Reply:
If no task is running, only the main thread might call it from `received` and if the main thread is currently processing `received`, it can not simultaneously be in any write operation, can it?

If on the other hand the task thread has written some response that is still in the buffer and then, upon finishing, decides to send the 100-continue of the next request on the channel, which is waiting for it, this might collide with `handle_write` of the main thread. So I guess the suggested change is really necessary.